### PR TITLE
[fix](fd) release fd for shutdown tablets

### PR DESCRIPTION
**Pull request by dataroaring**
When a tablet is shutdown, maybe there is a running compaction on it, the compaction would put fd in cache, so we should clear them when move the directory to the trash and fail the compaction due to shutdown state.

e.g.
t0: a compaction is running on tablet t;
t1: tablet t is set shutdown, files of tablet are removed from segment cache.
t2: the compaction stops and files of tablet are added to the segment cache again.
t3: tablet t is moved to trash due to it is not used any more.
t4: the files are deleted from trash.
t5: the tablet's cache in segment cache remains and the capacity consumed by the files is not freed.

When a stale rowset is expired, it is removed when it is not used any more, and remove() function deletes files and removes the cache. However, there is no harm in clearing cache when a rowset is expired.

We reconstruct code related to clearing cache in object oriented way.

## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by dataroaring** [comment]
run buildall

**Comment by dataroaring** [comment]
run buildall

**Comment by dataroaring** [comment]
run buildall

**Comment by dataroaring** [comment]
run buildall

File: be/src/olap/rowset/rowset.h [hunk 1]
```diff
@@ -303,6 +303,9 @@ class Rowset : public std::enable_shared_from_this<Rowset> {
 
     std::string get_rowset_info_str();
 
+    virtual void clear_inverted_index_cache() { LOG(INFO) << "should not reach here"; }
```
Comment (gavinchou) [inline be/src/olap/rowset/rowset.h:306]: LOG(ERROR)?

Comment (dataroaring) [reply be/src/olap/rowset/rowset.h:306]: error is horrible for users.

File: be/src/olap/rowset/beta_rowset.cpp [hunk 2]
```diff
@@ -124,6 +125,21 @@ Status BetaRowset::get_inverted_index_size(size_t* index_size) {
     return Status::OK();
 }
 
+void BetaRowset::clear_inverted_index_cache() {
```
Comment (Lchangliang) [inline be/src/olap/rowset/beta_rowset.cpp:128]: Is there any difference with BetaRowset::remove()?

**Comment by dataroaring** [comment]
run buildall

File: be/src/olap/rowset/beta_rowset.cpp (hunk 2 above)
Comment (dataroaring) [reply be/src/olap/rowset/beta_rowset.cpp:128]: remove() delets files too. When a tablet is moved to trash, the files are not deleted immediately. I let remove call clear_cache to reduce duplicated code.

**Comment by dataroaring** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

**Review by Lchangliang — APPROVED** [review]

**Comment by Lchangliang** [comment]
LGTM

File: be/src/olap/base_tablet.h [hunk 3]
```diff
@@ -241,6 +241,8 @@ class BaseTablet {
 
     std::vector<RowsetSharedPtr> get_snapshot_rowset(bool include_stale_rowset = false) const;
 
+    virtual void clear_cache() { LOG(INFO) << "should not reach here"; }
```
Comment (platoneko) [inline be/src/olap/base_tablet.h:244]:
```suggestion
    virtual void clear_cache() = 0;
```

File: be/src/olap/rowset/rowset.h (hunk 1 above)
Comment (platoneko) [reply be/src/olap/rowset/rowset.h:306]: Move this method to `BetaRowset` private field

File: be/src/olap/tablet.cpp
Comment (platoneko) [inline be/src/olap/tablet.cpp:1]:
```c++
void CloudTablet::clear_cache() override {
    CloudTablet::recycle_cached_data(get_snapshot_rowset(true));
    _engine.tablet_mgr().erase_tablet(tablet_id());
}

void CloudTablet::recycle_cached_data(const std::vector<RowsetSharedPtr>& rowsets) {
    for (auto& rs : rowsets) {
        // Clear cached opened segments and inverted index cache in memory
        rs->clear_cache();
    }

    if (config::enable_file_cache) {
        for (const auto& rs : rowsets) {
            for (int seg_id = 0; seg_id < rs->num_segments(); ++seg_id) {
                auto seg_path = rs->segment_file_path(seg_id);
                auto file_key = io::BlockFileCache::hash(io::Path(seg_path).filename().native());
                auto* file_cache = io::FileCacheFactory::instance()->get_by_path(file_key);
                file_cache->remove_if_cached(file_key);
            }
        }
    }
}
```

**Review by platoneko — CHANGES_REQUESTED** [review]

**Review by platoneko — DISMISSED** [review]

**Comment by dataroaring** [comment]
run buildall

**Comment by dataroaring** [comment]
run buildall

**Comment by dataroaring** [comment]
run buildall

**Comment by dataroaring** [comment]
run buildall

**Review by platoneko — APPROVED** [review]

**Pull request merged**